### PR TITLE
storageccl: Don't disable data key rotation after restart

### LINKED ISSUE #79066: storageccl: Encryption-at-rest data key rotation gets disabled after a restart
**Describe the problem**

When using encryption at rest, we need to be able to determine when the store and data keys have been rotated last.

**To Reproduce**

Encrypt a cluster.
Set a fast rotation-period
See that the "Created" time does not match the Store Key
See that the "Created" time for the Data Key, is in fact the Store Key.

**Expected behavior**
The "Created" times to match the created Store Key time and Data Key time.

**Additional data / screenshots**
image
^Notice how active Store key is 2022, whilst our automatically generated Data key is 2021

**Environment:**
 - CockroachDB version [21.1.12]
 - Linux

Jira issue: CRDB-14528

gz#11893

### PR DESCRIPTION
**Pull request by itsbilal**
Currently, there's a bug in encryption-at-rest FS initialization
code that unintentionally disables automaticdata key rotation unless
the store key was updated since the last start of the node.
This change fixes that.

Fixes #79066

Release note (bug fix): Address issue where automatic encryption-at-rest
data key rotation would get disabled after a node restart without
a store key rotation.

### CODE REVIEW

File: pkg/ccl/storageccl/engineccl/pebble_key_manager.go [hunk 1]
```diff
@@ -304,13 +304,14 @@ func (m *DataKeyManager) SetActiveStoreKeyInfo(
...
 		// so no need to do anything.
+		m.mu.rotationEnabled = true
 		return nil
 	}
 	// For keys other than plaintext, make sure the user is not reusing inactive keys.
 	if storeKeyInfo.EncryptionType != enginepbccl.EncryptionType_Plaintext {
 		if _, found := m.mu.keyRegistry.StoreKeys[storeKeyInfo.KeyId]; found {
 			return fmt.Errorf("new active store key ID %s already exists as an inactive key -- this"+
-				"is really dangerous", storeKeyInfo.KeyId)
+				"is really dangerous. data key rotation has been disabled.", storeKeyInfo.KeyId)
```
Comment (itsbilal) [inline pkg/ccl/storageccl/engineccl/pebble_key_manager.go:314]: Maybe this isn't the right call, and we should be enabling data key rotation in this case too. Thoughts?

Comment (itsbilal) [reply pkg/ccl/storageccl/engineccl/pebble_key_manager.go]: Nevermind, this case disables EAR.

**Review by nicktrav — COMMENTED** [review]
This seems like something we should have a test for.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-xlV2q9iO1oAIgQ4J9:b-jsqcgu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @shermanCRL, and @sumeerbhola)

___
*[-- commits, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-xkmCLEIYSGSTbfNB1):*
s/automaticdata/automatic data/

**Review by itsbilal — COMMENTED** [review]
Test added. TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-xxiKk204zrnk7L2VJ:b7j6360)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @shermanCRL and @sumeerbhola)

___
*[-- commits, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-xkmCLEIYSGSTbfNB1):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

s/automaticdata/automatic data/
</blockquote></details>

Done.

**Review by nicktrav — APPROVED** [review]
:lgtm: - I'm less familiar with this code, so worth getting another set of :eyes: on this.

Thanks for digging into it!

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-y3cgu8zNZnRTSVfzQ:b-gw9l4u)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @shermanCRL, and @sumeerbhola)

___
*[pkg/ccl/storageccl/engineccl/pebble_key_manager.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-y33Nd60Te4R6qS8QU:-N-y33Nd60Te4R6qS8QV:b-vhw0s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f21182e4a95aef84bb6d22309074f607d731f79/pkg/ccl/storageccl/engineccl/pebble_key_manager.go#L307)):*
> no need to do anything

nit: the bug here was that we weren't doing anything. Worth updating the comment to indicate that even though we're not rotating the key, it's important to keep rotation enabled?

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-y5lqXEDhK9BqFtiSc:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nicktrav, @shermanCRL, and @sumeerbhola)

___
*[pkg/ccl/storageccl/engineccl/pebble_key_manager.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-y33Nd60Te4R6qS8QU:-N-y5myu3vwp0HtdFkqY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f21182e4a95aef84bb6d22309074f607d731f79/pkg/ccl/storageccl/engineccl/pebble_key_manager.go#L307)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

> no need to do anything

nit: the bug here was that we weren't doing anything. Worth updating the comment to indicate that even though we're not rotating the key, it's important to keep rotation enabled?
</blockquote></details>

Done.

**Review by nicktrav — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @shermanCRL and @sumeerbhola)

**Review by jbowens — APPROVED** [review]
:lgtm:

Nice job tracking that down

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-y7FtCD1oT3JTxBLsm:b-3t26s9)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @shermanCRL and @sumeerbhola)

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @shermanCRL)

___
*[pkg/ccl/storageccl/engineccl/pebble_key_manager.go, line 302 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-yFRgbC45rlCNf5xEE:-N-yFRgbC45rlCNf5xEF:b-nfln4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/40188a16d845e8245ab882397f2e46de2ded79e5/pkg/ccl/storageccl/engineccl/pebble_key_manager.go#L302)):*
> ```Go
> 	}
> 	m.mu.Lock()
> 	defer m.mu.Unlock()
> ```

Can the rotationEnabled=true be hoisted to immediately after this defer?
I think this bug existed because of an oversight that this function could return early in a non-error case. And if we add more such paths in the future we may re-introduce this bug.

**Review by itsbilal — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N-yPbViCcPaSGR5zYHC:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jbowens, @nicktrav, @shermanCRL, and @sumeerbhola)

___
*[pkg/ccl/storageccl/engineccl/pebble_key_manager.go, line 302 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-N-yFRgbC45rlCNf5xEE:-N-yP_tgBgAcTXqnZWc9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40188a16d845e8245ab882397f2e46de2ded79e5/pkg/ccl/storageccl/engineccl/pebble_key_manager.go#L302)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can the rotationEnabled=true be hoisted to immediately after this defer?
I think this bug existed because of an oversight that this function could return early in a non-error case. And if we add more such paths in the future we may re-introduce this bug.
</blockquote></details>

Done.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80114#-:-N01d0uj68L-UrQOZDhP:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @shermanCRL and @sumeerbhola)

**Comment by itsbilal** [comment]
TFTRs!

bors r=sumeerbhola

**Pull request merged**